[bug]WebAuth Doesn't Work for testing images

### Prerequisites

- [x] I have searched the existing **Closed _AND_ Open** [Issues](https://github.com/dani-garcia/vaultwarden/issues?q=is%3Aissue%20) **_AND_** [Discussions](https://github.com/dani-garcia/vaultwarden/discussions?discussions_q=)
- [x] I have searched and read the [documentation](https://github.com/dani-garcia/vaultwarden/wiki/)

### Vaultwarden Support String

### Your environment (Generated via diagnostics page)

* Vaultwarden version: v1.34.3-a133d4e9
* Web-vault version: v2025.7.0
* OS/Arch: linux/x86_64
* Running within a container: true (Base: Debian)
* Database type: SQLite
* Database version: 3.50.2
* Uses config.json: false
* Uses a reverse proxy: true
* IP Header check: true (X-Forwarded-For)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* TZ environment: Asia/Shanghai
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Websocket Check: true
* HTTP Response Checks: true

### Config & Details (Generated via diagnostics page)

<details><summary>Show Config & Details</summary>

**Config:**

```json
{
  "_duo_akey": null,
  "_enable_duo": true,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_max_note_size": 10000,
  "_smtp_img_src": "***:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_connect_src": "",
  "allowed_iframe_ancestors": "*****://***********",
  "attachments_folder": "data/attachments",
  "auth_request_purge_schedule": "30 * * * * *",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://************",
  "domain_origin": "*****://************",
  "domain_path": "",
  "domain_set": true,
  "duo_context_purge_schedule": "30 * * * * *",
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "duo_use_iframe": false,
  "email_2fa_auto_fallback": false,
  "email_2fa_enforce_on_verified_invite": false,
  "email_attempts_limit": 3,
  "email_change_allowed": false,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "enable_websocket": true,
  "enforce_single_org_with_reset_pw_policy": false,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "experimental_client_feature_flags": "autofill-overlay,autofill-v2,browser-fileless-import,extension-refresh,fido2-vault-credentials,inline-menu-positioning-improvements,ssh-key-vault-item,ssh-agent,export-attachments,anon-addy-self-host-alias,simple-login-self-host-alias,mutual-tls",
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": "***",
  "http_request_block_non_global_ips": true,
  "http_request_block_regex": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "increase_note_size_limit": false,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "KyzDt",
  "invitations_allowed": true,
  "ip_header": "X-Forwarded-For",
  "job_poll_interval_ms": 30000,
  "log_file": "/log/vaultwarden.log",
  "log_level": "debug",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f%z",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": true,
  "password_hints_allowed": true,
  "password_iterations": 600000,
  "purge_incomplete_sso_nonce": "0 20 0 * * *",
  "push_enabled": true,
  "push_identity_uri": "https://identity.bitwarden.eu",
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://api.bitwarden.eu",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "",
  "signups_verify": true,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "***************",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "**************",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "***************",
  "sso_allow_unknown_email_verification": false,
  "sso_audience_trusted": null,
  "sso_auth_only_not_session": false,
  "sso_authority": "https://xxxxx",
  "sso_authorize_extra_params": "",
  "sso_callback_path": "https://xxxxx/identity/connect/oidc-signin",
  "sso_client_cache_expiration": 0,
  "sso_client_id": "xxxx",
  "sso_client_secret": "***",
  "sso_debug_tokens": true,
  "sso_enabled": false,
  "sso_master_password_policy": null,
  "sso_only": false,
  "sso_pkce": true,
  "sso_scopes": "openid email profile offline_access",
  "sso_signups_match_email": true,
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "user_send_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```

</details>


### Vaultwarden Build Version

testing

### Deployment method

Official Container Image

### Custom deployment method

_No response_

### Reverse Proxy

nginx

### Host/Server Operating System

Linux

### Operating System Version

Mac

### Clients

Browser Extension

### Client Version

_No response_

### Steps To Reproduce

1. I try to configure OIDC, but it doesn't work
2. Then I set `SSO_ENABLED = false` and use master password to login
3. It asked me to provide webauth, it passed but not action

### Expected Result

should login

### Actual Result

no action after webauth

### Logs

```text
vaultwarden  | [2025-08-09 10:08:27.016+0800][start][INFO] Rocket has launched from https://0.0.0.0:80
vaultwarden  | [2025-08-09 10:08:31.249+0800][rustls::server::hs][DEBUG] decided upon suite TLS13_AES_256_GCM_SHA384
vaultwarden  | [2025-08-09 10:08:31.252+0800][request][INFO] POST /identity/accounts/prelogin
vaultwarden  | [2025-08-09 10:08:31.253+0800][response][INFO] (prelogin) POST /identity/accounts/prelogin => 200 OK
vaultwarden  | [2025-08-09 10:08:31.258+0800][rustls::common_state][DEBUG] Sending warning alert CloseNotify
vaultwarden  | [2025-08-09 10:08:31.628+0800][rustls::server::hs][DEBUG] decided upon suite TLS13_AES_256_GCM_SHA384
vaultwarden  | [2025-08-09 10:08:31.630+0800][request][INFO] POST /identity/connect/token
vaultwarden  | [2025-08-09 10:08:31.957+0800][error][ERROR] 2FA token not provided
vaultwarden  | [2025-08-09 10:08:31.957+0800][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
vaultwarden  | [2025-08-09 10:08:31.958+0800][rustls::common_state][DEBUG] Sending warning alert CloseNotify
vaultwarden  | [2025-08-09 10:08:33.609+0800][rustls::server::hs][DEBUG] decided upon suite TLS13_AES_256_GCM_SHA384
vaultwarden  | [2025-08-09 10:08:33.611+0800][request][INFO] GET /webauthn-fallback-connector.html?data=xxxxx
vaultwarden  | [2025-08-09 10:08:33.611+0800][response][INFO] (web_files) GET /<p..> [10] => 200 OK
vaultwarden  | [2025-08-09 10:08:33.613+0800][rustls::common_state][DEBUG] Sending warning alert CloseNotify
vaultwarden  | [2025-08-09 10:08:56.735+0800][vaultwarden::api::core::two_factor][DEBUG] Sending notifications for incomplete 2FA logins
vaultwarden  | [2025-08-09 10:08:56.735+0800][vaultwarden::api::core::two_factor::duo_oidc][DEBUG] Purging Duo authentication contexts
vaultwarden  | [2025-08-09 10:08:56.735+0800][vaultwarden::api::core::accounts][DEBUG] Purging auth requests
vaultwarden  | [2025-08-09 10:09:06.724+0800][rustls::server::hs][DEBUG] decided upon suite TLS13_AES_256_GCM_SHA384
vaultwarden  | [2025-08-09 10:09:06.725+0800][request][INFO] GET /webauthn-fallback-connector.html?data=eyJhbGxvd0NyZWRlbnRpYWxzI
vaultwarden  | [2025-08-09 10:09:06.726+0800][response][INFO] (web_files) GET /<p..> [10] => 200 OK
vaultwarden  | [2025-08-09 10:09:06.728+0800][rustls::common_state][DEBUG] Sending warning alert CloseNotify
vaultwarden  | [2025-08-09 10:09:26.584+0800][rustls::server::hs][DEBUG] decided upon suite TLS13_AES_256_GCM_SHA384
vaultwarden  | [2025-08-09 10:09:26.584+0800][rustls::server::hs][DEBUG] Chosen ALPN protocol [xxxx]
vaultwarden  | [2025-08-09 10:09:26.588+0800][request][INFO] GET /alive
vaultwarden  | [2025-08-09 10:09:26.589+0800][response][INFO] (alive) GET /alive => 200 OK
vaultwarden  | [2025-08-09 10:09:26.591+0800][rustls::common_state][DEBUG] Sending warning alert CloseNotify
vaultwarden  | [2025-08-09 10:09:56.737+0800][vaultwarden::api::core::two_factor][DEBUG] Sending notifications for incomplete 2FA logins
vaultwarden  | [2025-08-09 10:09:56.737+0800][vaultwarden::api::core::accounts][DEBUG] Purging auth requests
vaultwarden  | [2025-08-09 10:09:56.737+0800][vaultwarden::api::core::two_factor::duo_oidc][DEBUG] Purging Duo authentication contexts
vaultwarden  | [2025-08-09 10:09:56.738+0800][vaultwarden::api::core::two_factor][INFO] User iamtaochen@outlook.com did not complete a 2FA login within the configured time limit. IP: xxxx
vaultwarden  | [2025-08-09 10:09:56.749+0800][rustls::webpki::anchors][DEBUG] add_parsable_certificates processed 142 valid and 0 invalid certs
vaultwarden  | [2025-08-09 10:09:58.246+0800][rustls::client::hs][DEBUG] No cached session for DnsName("xxxx")
vaultwarden  | [2025-08-09 10:09:58.246+0800][rustls::client::hs][DEBUG] Not resuming any session
vaultwarden  | [2025-08-09 10:09:58.613+0800][rustls::client::hs][DEBUG] Using ciphersuite TLS13_AES_256_GCM_SHA384
vaultwarden  | [2025-08-09 10:09:58.613+0800][rustls::client::tls13][DEBUG] Not resuming
vaultwarden  | [2025-08-09 10:09:58.614+0800][rustls::client::tls13][DEBUG] TLS1.3 encrypted extensions: ServerExtensions { unknown_extensions: {}, .. }
vaultwarden  | [2025-08-09 10:09:58.615+0800][rustls::client::hs][DEBUG] ALPN protocol is None
vaultwarden  | [2025-08-09 10:10:02.949+0800][rustls::common_state][DEBUG] Sending warning alert CloseNotify
vaultwarden  | [2025-08-09 10:10:26.690+0800][rustls::server::hs][DEBUG] decided upon suite TLS13_AES_256_GCM_SHA384
vaultwarden  | [2025-08-09 10:10:26.690+0800][rustls::server::hs][DEBUG] Chosen ALPN protocol [1xxxx]
vaultwarden  | [2025-08-09 10:10:26.694+0800][request][INFO] GET /alive
vaultwarden  | [2025-08-09 10:10:26.694+0800][response][INFO] (alive) GET /alive => 200 OK
vaultwarden  | [2025-08-09 10:10:26.696+0800][rustls::common_state][DEBUG] Sending warning alert CloseNotify
vaultwarden  | [2025-08-09 10:10:56.739+0800][vaultwarden::api::core::two_factor][DEBUG] Sending notifications for incomplete 2FA logins
vaultwarden  | [2025-08-09 10:10:56.740+0800][vaultwarden::api::core::accounts][DEBUG] Purging auth requests
vaultwarden  | [2025-08-09 10:10:56.740+0800][vaultwarden::api::core::two_factor::duo_oidc][DEBUG] Purging Duo authentication contexts
```

### Screenshots or Videos

_No response_

### Additional Context

_No response_

@Blacks-Army, i would suggest to fix those items, as they might cause some weird issues with some clients.

I tried on a macOS device, and I'm able to add a key, but afterwards I'm unable to login with it.
```
[2025-08-09 19:15:42.416][webauthn_rs_core::core][DEBUG] extensions: AuthenticationSignedExtensions { unknown_keys: {} }
[2025-08-09 19:15:42.416][vaultwarden::api::core::two_factor::webauthn][ERROR] Credential not present
```

Using my YubiKey5c on the mac works fine though.

--- a/src/api/core/two_factor/webauthn.rs
+++ b/src/api/core/two_factor/webauthn.rs
@@ -4,6 +4,7 @@ use crate::{
         EmptyResult, JsonResult, PasswordOrOtpData,
     },
     auth::Headers,
+    crypto::ct_eq,
     db::{
         models::{EventType, TwoFactor, TwoFactorType, UserId},
         DbConn,
@@ -434,12 +435,14 @@ pub async fn validate_webauthn_login(
     let authentication_result = webauthn.finish_securitykey_authentication(&rsp, &state)?;
 
     for reg in &mut registrations {
-        if reg.credential.cred_id() == authentication_result.cred_id() && authentication_result.needs_update() {
-            reg.credential.update_credential(&authentication_result);
-
-            TwoFactor::new(user_id.clone(), TwoFactorType::Webauthn, serde_json::to_string(&registrations)?)
-                .save(conn)
-                .await?;
+        if ct_eq(reg.credential.cred_id(), authentication_result.cred_id()) {
+            // If the cred id matches and the credential is updated, Some(true) is returned
+            // In those cases, update the record, else leave it alone
+            if reg.credential.update_credential(&authentication_result) == Some(true) {
+                TwoFactor::new(user_id.clone(), TwoFactorType::Webauthn, serde_json::to_string(&registrations)?)
+                    .save(conn)
+                    .await?;
+            }
             return Ok(());
         }
     }

